controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1460,32 +1487,34 @@ private Segment transform(StreamSegmentRecord segmentRecord) {
         return segmentRecords.stream().map(this::transform).collect(Collectors.toSet());
     }
     
-    private CompletableFuture<List<EpochRecord>> fetchEpochs(int fromEpoch, int toEpoch) {
+    @VisibleForTesting
+    CompletableFuture<List<EpochRecord>> fetchEpochs(int fromEpoch, int toEpoch, boolean ignoreCache) {
         // fetch history time series chunk corresponding to from.
         // read entries till either last entry or till to
         // if to is not in this chunk fetch the next chunk and read till to
         // keep doing this until all records till to have been read.
         // keep computing history record from history time series by applying delta on previous.
-        return getActiveEpochRecord(false)
-                .thenApply(currentEpoch -> currentEpoch.getEpoch() / HISTORY_CHUNK_SIZE)
+        return getActiveEpochRecord(ignoreCache)
+                .thenApply(currentEpoch -> currentEpoch.getEpoch() / historyChunkSize.get())
                 .thenCompose(latestChunkNumber -> Futures.allOfWithResults(
-                        IntStream.range(fromEpoch / HISTORY_CHUNK_SIZE, toEpoch / HISTORY_CHUNK_SIZE + 1)
+                        IntStream.range(fromEpoch / historyChunkSize.get(), toEpoch / historyChunkSize.get() + 1)
                                  .mapToObj(i -> {
-                                     int firstEpoch = i * HISTORY_CHUNK_SIZE > fromEpoch ? i * HISTORY_CHUNK_SIZE : fromEpoch;
+                                     int firstEpoch = i * historyChunkSize.get() > fromEpoch ? i * historyChunkSize.get() : fromEpoch;
+
                                      boolean ignoreCached = i >= latestChunkNumber;
-                                     return getEpochsFromHistoryChunk(i, fromEpoch, toEpoch, firstEpoch, ignoreCached);
+                                     return getEpochsFromHistoryChunk(i, firstEpoch, toEpoch, ignoreCached);
                                  }).collect(Collectors.toList())))
                 .thenApply(c -> c.stream().flatMap(Collection::stream).collect(Collectors.toList()));
     }
 
-    private CompletableFuture<List<EpochRecord>> getEpochsFromHistoryChunk(int chunk, int fromEpoch, int toEpoch, int firstEpoch, boolean ignoreCached) {
+    private CompletableFuture<List<EpochRecord>> getEpochsFromHistoryChunk(int chunk, int firstEpoch, int toEpoch, boolean ignoreCached) {
```

Comment:
I don't understand the change from `from` to `first`. Actually, the previous situation was even more confusing as we had `first`, `from`, and `to`. Can you elaborate on the semantics of the arguments here?

Reply:
so the logic is as follows:
`from` and `to` are supplied by caller. 
They both can lie in different chunks and from is not necessarily the starting entry in the chunk it is found in. 

Using `from` and `to` we can compute the chunk number (modulo operation). 
So we call all the chunks in parallel and ask them to read all epochs between `from` and `to`. 

`first` simply tells this method which is the `first` entry to be read from in the chunk. 
For chunk which contains `from` the `first` is equal to from. For every other chunk the `first` corresponds to first record in the chunk. 

Since HistoryTimeSeries stores epoch deltas, we send the `first` as an optimization that read full record for `first` and then apply epoch deltas in the chunk to compute full entries for all intermediate epochs between from and to.